Add support for encoding X500DistinguishedName values on Unix.

bartonjs opened the pull request:
The Unix implementation matches the Windows behavior on the Use*
flags for choosing the RDN separator, but always uses the UTF-8
encoding instead of the UCS-2/BMP encoding (Windows will use UTF-8 as
an opt-in with a flag).

So new X500DistinguishedName(input, flags).Decode(false) will always
produce the same output string from both Windows and Unix, but the RawData
will differ.

Fixes #1993. (This is the last sub-issue)

cc: @stephentoub, @AtsushiKan, @morganbr

bartonjs commented:
Pushed a new commit; I'll squash after signoff.

morganbr commented on src/Common/src/System/Security/Cryptography/DerEncoder.cs:
```diff
@@ -445,6 +445,74 @@ internal static byte[][] SegmentedEncodeOid(Oid oid)
+            //
+            // Thankfully, that's just Encoding.UTF8.
+
+            unsafe
```
I'm not a fan of unsafe code when there's a safe equivalent and there's not an important reason for it. Can you switch to the char[] overload and just use safe code here?

bartonjs replied:
Switched almost everything to char\[\], and found a couple bad comments while doing it.

bartonjs commented:
Apparently I introduced a data error in the adjacent quotes test after the last time I ran it on my Linux box.  Fixed that, so it should rerun/pass now.

@AtsushiKan @morganbr Any other comments?

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/X500NameEncoder.cs:
```diff
@@ -26,6 +30,11 @@ internal static class X500NameEncoder
+        private static readonly List<char> s_useSemicolonSeparators = new List<char>(1) { ';' };
+        private static readonly List<char> s_useCommaSeparators = new List<char>(1) { ',' };
+        private static readonly List<char> s_useNewlineSeparators = new List<char>(2) { '\r', '\n' };
+        private static readonly List<char> s_defaultSeparators = new List<char>(2) { ',', ';' };
```
Why ```List<char>``` rather than ```char[]```?

bartonjs replied:
Mainly just that I thought that calling `separators.Contains(c)` looked a lot better than `separators.IndexOf(c) >= 0`; and went with "reads better" since there's only ever the 4.

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/X500NameEncoder.cs:
```diff
@@ -200,5 +252,445 @@ private static void AppendOid(StringBuilder decodedName, string oidValue)
+            // 7 + 6 = 13, round up to the nearest power-of-two.
+            const int InitalRdnSize = 16;
+            List<byte[][]> encodedSets = new List<byte[][]>(InitalRdnSize);
+            char[] chars = stringForm.ToCharArray();
```
Why is the ToCharArray needed, rather than just indexing into stringForm?

bartonjs replied:
The two options for encoding a substring (via System.Text.Encoding) are `char*` or `char[]`, since GetByteCount doesn't accept an offset+length for a string, and the GetBytes that does requires a sufficiently sized buffer as input.

stephentoub replied:
I realize there's lots of other allocation happening, but my personal preference would be to avoid the extra allocation/copy and use unsafe with GetByteCount and ```char*``` rather than creating the ```char[]```.  I saw Morgan's comments about not using unsafe when there's no value, but here there seems to be value.

stephentoub commented on src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/X500NameEncoder.cs:
```diff
@@ -200,5 +252,445 @@ private static void AppendOid(StringBuilder decodedName, string oidValue)
+            bool hadEscapedQuote = false;
+
+            const char keyValueSeparator = '=';
+            const char quotedValueChar = '"';
```
Nit: consts should begin with a capital

bartonjs replied:
Fixed.

stephentoub commented:
LGTM

bartonjs commented:
Squashed down to a single commit.  In case anyone still wants to give late feedback, the individual commits are still available at https://github.com/bartonjs/corefx/commits/unix_x500dn_encode_reviewed

bartonjs commented:
Test Innerloop Windows_NT Release Build and Test please
